src/Controller/Api/UploadController.php

```diff
@@ -2,98 +2,113 @@
+     * @param EntityManagerInterface $entityManager Entity manager interface instance.
      *
-     * @var uploadHandler
-     */
-    protected $entityHandler;
-
-    /**
-     * The form factory.
+     * @View()
      *
-     * @var FormFactoryInterface Form factory instance.
-     */
-    protected $formFactory;
-
-    /**
-     * TreeController constructor.
+     * @Route(
+     *     "/api/files/upload-chunks",
+     *     name="pelagos_api_post_chunks",
+     *     methods={"POST"},
+     *     defaults={"_format"="json"}
+     *     )
      *
-     * @param UploadHandler        $uploadHandler The upload handler.
-     * @param EntityHandler        $entityHandler The entity handler.
-     * @param FormFactoryInterface $formFactory   The form factory.
+     * @return Response The result of the post.
      */
-    public function __construct(UploadHandler $uploadHandler, EntityHandler $entityHandler, FormFactoryInterface $formFactory)
+    public function postChunks(Request $request, FileUploader $fileUploader, EntityManagerInterface $entityManager)
     {
-        $this->entityHandler = $entityHandler;
-        $this->formFactory = $formFactory;
-        parent::__construct($entityHandler, $formFactory);
-        $this->uploadHandler = $uploadHandler;
+        $this->denyAccessUnlessGranted('IS_AUTHENTICATED_FULLY');
+        try {
+            $fileUploader->uploadChunk($request);
+        } catch (\Exception $exception) {
+            throw new BadRequestHttpException($exception->getMessage());
+        }
+
+        return $this->makeNoContentResponse();
     }
 
     /**
-     * Process a post of a file or a file chunk.
+     * Combine file chunks.
      *
-     * @param Request $request The Symfony request object.
+     * @param Request                $request       The Symfony request object.
+     * @param FileUploader           $fileUploader  File upload handler service.
+     * @param EntityManagerInterface $entityManager Entity manager interface instance.
+     * @param string                 $id            Dataset submission id.
      *
      * @View()
      *
      * @Route(
-     *     "/api/upload",
-     *     name="pelagos_api_upload_post",
+     *     "/api/files/combine-chunks/{id}",
+     *     name="pelagos_api_combine_chunks",
      *     methods={"POST"},
      *     defaults={"_format"="json"}
      *     )
      *
-     * @return array The result of the post.
+     * @return Response The result of the post.
      */
-    public function postAction(Request $request)
+    public function combineChunks(Request $request, FileUploader $fileUploader, EntityManagerInterface $entityManager, string $id)
     {
         $this->denyAccessUnlessGranted('IS_AUTHENTICATED_FULLY');
-        return $this->uploadHandler->handleUpload($request);
+        try {
+            $fileMetadata = $fileUploader->combineChunks($request);
+            $this->updateFileEntity($fileMetadata, $entityManager, $id);
```

Comment:
I still don't think combineChunks should be responsible for updating the entity, it should just return the path where the file was uploaded. Then the combineChunks event should tell the submission api to add a file. We could do this in a later PR.